### ISSUE TITLE
Stumped by multilingual relation traversal

### ISSUE THREAD

**Issue by francis-dion** [goodmami/wn#168]
I can't figure out how to reliably reach the expanded "Inferred" relations for a given synset.

Below is the code I'm using followed by  the results I get for synset omw-fr-13861050-n.
omw-fr gives me 12 "inferred" related synsets, but the translation through 'oewn:2021' gives me only 8 related, 6 of which I can't translate back to French.

What am I missing here?

Thanks for your help!

```
import wn
expandingLexicon = 'oewn' #omw-en:1.4 or oewn
mywn = wn.Wordnet('omw-fr', expand=expandingLexicon)
form = 'plan'
word = mywn.words(form)[2]
synset = word.synsets()[0]
translation = synset.translate(expandingLexicon)[0]
print('Words:', len(mywn.words(form)))
print('Word.Synsets:', len(word.synsets()))
print('Synset.Related:', len(synset.get_related()))
print('*** synset.Inferred:', len([s for s in synset.get_related() if str(s) == 'Synset(\'*INFERRED*\')']))
print('synset.Translations:', len(synset.translate(expandingLexicon)))
print('*** translation.Related:', len(translation.get_related()))
print('The Related for', synset, synset.lemmas())
print(*[(ss, ss.lemmas()) for ss in synset.get_related()], sep='\n')
print('The translation.Related for', translation, translation.lemmas() )
print(*[(ss, ss.lemmas()) for ss in translation.get_related()], sep='\n')
print('The Translations of the Translated Related')
print(*[([] if mywn.synsets(ili=ss.ili.id) == [] else (mywn.synsets(ili=ss.ili.id)[0], mywn.synsets(ili=ss.ili.id)[0].lemmas())) for ss in translation.get_related()], sep='\n')
```

```
Words: 3
Word.Synsets: 11
Synset.Related: 14
*** synset.Inferred: 12
synset.Translations: 1
*** translation.Related: 8
The Related for Synset('omw-fr-13861050-n') ['plan']
(Synset('omw-fr-00027807-n'), ['forme'])
(Synset('omw-fr-06000644-n'), ['maths', 'mathématiques'])
(Synset('*INFERRED*'), [])
(Synset('*INFERRED*'), [])
(Synset('*INFERRED*'), [])
(Synset('*INFERRED*'), [])
(Synset('*INFERRED*'), [])
(Synset('*INFERRED*'), [])
(Synset('*INFERRED*'), [])
(Synset('*INFERRED*'), [])
(Synset('*INFERRED*'), [])
(Synset('*INFERRED*'), [])
(Synset('*INFERRED*'), [])
(Synset('*INFERRED*'), [])
The translation.Related for Synset('oewn-13883265-n') ['plane', 'sheet']
(Synset('oewn-00028005-n'), ['form', 'shape'])
(Synset('oewn-06009822-n'), ['mathematics', 'maths', 'math'])
(Synset('oewn-13883664-n'), ['Cartesian plane'])
(Synset('oewn-13884183-n'), ['facet plane'])
(Synset('oewn-13884289-n'), ['midplane', 'midline'])
(Synset('oewn-13884409-n'), ['orbital plane'])
(Synset('oewn-13884537-n'), ['picture plane'])
(Synset('oewn-13884747-n'), ['tangent plane'])
The Translations of the Translated Related
(Synset('omw-fr-00027807-n'), ['forme'])
(Synset('omw-fr-06000644-n'), ['maths', 'mathématiques'])
[]
[]
[]
[]
[]
[]
```

**Comment by goodmami** [comment]
To make sure we're on the same page, the INFERRED synsets are those that *do not* exist in the lexicon, but their existence/absence is inferred because of a relation borrowed from another lexicon. So:

* Lexicon `A` relies on the "expand" lexicon `B`
* Synset `b1` in lexicon `B` has a relation `r1-2` to synset `b2`
* Synset `a1` exists in lexicon `A` with the same ILI as `b1`, but there is no synset with the same ILI as `b2`
* When traversing relations from `a1`, `r1-2` is borrowed but lexicon `A` cannot resolve its target with a synset, so it is an INFERRED synset with the same ILI as `b2`

Therefore, if you try to translate from `b2` into lexicon `A`, you will not find anything because there is no synset with the same ILI.

By the way, when I follow your code I see 8 related synsets, not 14:

```pycon
>>> import wn
>>> fr = wn.Wordnet('omw-fr', expand='oewn')
>>> len(fr.synsets('plan', pos='n'))
11
>>> synset = fr.synsets('plan', pos='n')[0]
>>> len(synset.get_related())
8
```

Getting 14 related synsets seems like the 6 INFERRED synsets were duplicated somehow, but the 2 non-inferred ones were left alone. This might happen if you have multiple expand lexicons covering the same ILIs:

```pycon
>>> fr = wn.Wordnet('omw-fr', expand='oewn omw-en')  # two expand lexicons
>>> synset = fr.synsets('plan', pos='n')[0]
>>> len(synset.get_related())
14
```

Hope this helps!

BTW, when investigating this I found a bug with the `Synset.relations()` method. See #169.

**Comment by francis-dion** [comment]
It does help, re: not finding the back-translations. Also good catch on the doubling of INFERRED, where, I suspect, there's either a bug or something else I don't get. :-)
My Wordnet object is created with just the one expand lexicon.
```
expandingLexicon = 'oewn'
mywn = wn.Wordnet('omw-fr', expand=expandingLexicon)
```
Which I confirmed:
```
print(mywn.expanded_lexicons())
## [<Lexicon oewn:2021 [en]>]
```
Now, omw-en is loaded in the database, but I thought the point of creating the Wordnet object was to control what expand lexicon was being used. All my objects (synset, word, translation) are derived (created from) mywn, but maybe this origin isn't sufficient. I see this for example:
```
print(word.lexicon().requires())
## {'omw-en:1.4': <Lexicon omw-en:1.4 [en]>}
```
Again, is there something I am missing?

**Comment by goodmami** [comment]
The `requires()` method lists which dependencies have been declared for the lexicon, not necessarily which will be used. See `Wordnet.expanded_lexicons()` to see which ones will be used. If `expand` is not specified when instantiating a `Wordnet` object, then the dependency lexicon will be automatically used as an expand lexicon if it is available on the system.

In this example, `omw-fr` specifies `omw-en:1.4` as its dependency and this lexicon has been installed to the database:

```pycon
>>> fr = wn.Wordnet('omw-fr')  # no 'expand' specified
>>> fr.expanded_lexicons()
[<Lexicon omw-en:1.4 [en]>]
>>> fr = wn.Wordnet('omw-fr', expand=None)  # same as not specifying
>>> fr.expanded_lexicons()
[<Lexicon omw-en:1.4 [en]>]
>>> fr = wn.Wordnet('omw-fr', expand='')  # explicitly set no expand lexicons
>>> fr.expanded_lexicons()
[]
>>> fr = wn.Wordnet('omw-fr', expand='oewn')  # set a different expand lexicon
>>> fr.expanded_lexicons()
[<Lexicon oewn:2021 [en]>]
```

If (1) you're certain you only have one expand lexicon set and (2) you're doing secondary queries from objects created from that lexicon and (3) you're still seeing the duplicated INFERRED synsets, then there might be a bug. I'm unable to reproduce this locally, so I may need to ask you to do some digging, in that case.

**Comment by goodmami** [comment]
Ok, thanks for isolating and identifying the problem, and good sleuthing :)

The fix you proposed may be useful in the general case but I think it's best to address the problem directly so it behaves the same as other, similar methods. To explain: objects of queries (words, senses, synsets) have the `_wordnet` member (as you've noticed) to keep track of the query parameters like `lexicon` and `expand` for secondary queries, but it is not being passed to synsets created from senses. (Also note that the `_wordnet` member is non-public and should not be used in regular code. It is only used internally in Wn's implementation, at least for now.)

For example, here is how `Sense` objects are created from a `Word`:

https://github.com/goodmami/wn/blob/a832a63321c20492a8be9f0bb3308ad3a684a516/wn/_core.py#L404-L406

Note how the `_wordnet` member is passed to the initializer. Now here is how a `Synset` is created from a `Sense`:

https://github.com/goodmami/wn/blob/a832a63321c20492a8be9f0bb3308ad3a684a516/wn/_core.py#L905

If you want to provide a fix, the second should look more like the first. The PR should include a test that fails before the test is applied. Here is such a test that might go in `tests/relation_tests.py`:

```python
@pytest.mark.usefixtures('mini_db_1_1')
def test_sense_synset_issue_168():
    ja = wn.Wordnet(lexicon='test-ja', expand='')
    assert ja.synset('test-ja-0001-n').get_related() == []
    assert ja.sense('test-ja-情報-n-0001-01').synset().get_related() == []
```

Feel free to submit a PR with the code you have and we can discuss how to refine it so it can be merged.

**Comment by francis-dion** [comment]
Ok I adjusted my fix to be more complete/generic, validated it with a test that failed before and worked after, and initiated the pull request.
Let me know if there's anything you'd like me to change / do differently.
FYI I also added a line to .gitignore to avoid importing files specific to my IDE (PyCharm).
Cheers!

**Comment by goodmami** [comment]
Closed in #170 (7479463f72e51d946c7e4af09cd59df9a9a22e20)
